Add post-mortem for incident on the 6th of May 2022 (#358)
openSUSE/obs-landing: _posts/deployments/2022-05-10-post-mortem.md

---
layout: post
title: "Post-Mortem: Conflict with ruby standard gems on May 10, 2022"
category: deployments
author: Lukas
---

There was a severe service degradation of our [reference server](https://build.opensuse.org). We want to give you some insight into what happened.

On May 10 we were running the deployment of OBS, which included an minor version update of the rubygem `strscan`. This lead to a downtime of 20 minutes.

## Impact

> build.opensuse.org was offline for 20 minutes

Reviewer: ```suggestion
build.opensuse.org was offline for 20 minutes and no one was able to work with the API or user interface. software.opensuse.org was taken offline because of this too.
```

---
layout: post

> title: "Post-Mortem: Conflict with ruby standard gems on May 10, 2022"

Reviewer 2: The incident was on May 10, ok. But we are going to publish the blog post today maybe. Can you change the date in the file name, please?  So `_posts/deployments/2022-05-11-post-mortem.md`

category: deployments
author: Lukas
---

There was a severe service degradation of our [reference server](https://build.opensuse.org). We want to give you some insight into what happened.

> On May 10 we were running the deployment of OBS, which included an minor version update of the rubygem `strscan`. This lead to a downtime of 20 minutes.

Reviewer 2: ```suggestion
On May 10 we were running the deployment of OBS, which included a minor version update of the rubygem `strscan`. This led to a downtime of 20 minutes.
```

## Impact

> build.opensuse.org was offline for 20 minutes

Reviewer 2: ```suggestion
build.opensuse.org was offline for 20 minutes.
```

## Root Causes

> After an minor update of the rubygem `strscan`, defined in the Gemfile of the Ruby on Rails project of OBS, there was a conflict with the `strscan` version provided by Ruby 3.1. Due to the conflict, Passenger wasn't able to spawn new instances of the application.

Reviewer: ```suggestion
After the update of the `strscan` ruby gem in our bundle, we had two `strscan` versions on our production installation. One provided by Ruby 3.1 stdlib (3.0.1) and one by our gem bundle (3.0.2). bundler itself requires `strscan` to function and loads from stdlib by default. During booting the application bundler noticed the mismatch and wasn't able to spawn new instances of the application.

```
Could not spawn process for application /srv/www/obs/api: The application encountered the following error: You have already activated strscan 3.0.1, but your Gemfile requires strscan 3.0.2. Since strscan is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports strscan as a default gem. (Gem::LoadError)
```
```

> After an minor update of the rubygem `strscan`, defined in the Gemfile of the Ruby on Rails project of OBS, there was a conflict with the `strscan` version provided by Ruby 3.1. Due to the conflict, Passenger wasn't able to spawn new instances of the application.

Reviewer 2: ```suggestion
After a minor update of the rubygem `strscan`, defined in the Gemfile of the Ruby on Rails project of OBS, there was a conflict with the `strscan` version provided by Ruby 3.1. Due to the conflict, Passenger wasn't able to spawn new instances of the application.
```

## Trigger

Merging and deploying the `strscan` gem update, created by Depfu.

## Detection

The deployment with Ansible showed failures and shortly after we received alerts in
our team chat.

## Resolution

Setting back the version of `strscan` by monkey patching the version number defined
in the `Gemfile.lock` and restarting the Apache server resolved the issue temporary.

> Reverting the version update upstream and deploying again resolved the issue permanently.

Reviewer: ```suggestion
Reverting the version update upstream and deploying again resolved the issue permanently.

In the meanwhile this has also been addressed in the bundler dependency [`psych`](https://github.com/ruby/psych/pull/534).
```

## Action Items

> - Pin the version of `strscan` to avoid conflicts in the future.

Reviewer: ```suggestion
- Pin the version of `strscan` until a `psych` version with the fix is released.
```

## Lessons Learned

### What Went Well

Collaboration among the team to resolve this.

### What Went Wrong

Missing awareness of the possible conflict between the gem versions defined in

> the rails project and the ones provided by the ruby core.

Reviewer 2: ```suggestion
the Rails project and the ones provided by the Ruby core.
```

### Where We Got Lucky

> No permanant damage or data loss.

Reviewer 2: ```suggestion
No permanent damage or data loss.
```

## Timeline (times in UTC)

- 14:24 Started the deployment of build.opensuse.org which wasn't successful.

> - 14:28 Detected outage: We received alerts in our team chat and build.opensuse.org wasn't accessible anymore.

Reviewer: 4 minutes? Really?

Author: yeah it was less, I realized right after the deployment, was more about contacting the team

Reviewer: It's not about the team and this sounds very wrong...

Author: No thats what not I meant, I started looking into it and then after 4 minutes I contacted the team. So talking about the delay, not the team taking long. I will change it

> - 14:32 Detected the conflict of the gem versions in the logfiles.

Reviewer 2: ```suggestion
- 14:32 Detected the conflict of the gem versions in the log files.
```

> - 14:43 Monkey patched the version of `strscan` from 3.0.2 back to 3.0.1 in the `Gemfile.lock` and restarted the Apache2 server

Reviewer 2: ```suggestion
- 14:43 Monkey patched the version of `strscan` from 3.0.2 back to 3.0.1 in the `Gemfile.lock` and restarted the Apache2 server.
```

> - 14:44 build.o.o was back and alerts were resolved

Reviewer 2: ```suggestion
- 14:44 build.o.o was back and alerts were resolved.
```
